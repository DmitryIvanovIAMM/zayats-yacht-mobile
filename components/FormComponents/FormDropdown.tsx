import { errorColor, secondary } from "@/constants/Colors";
import React, { useImperativeHandle, useRef } from "react";
import { Platform, StyleSheet, Text, View } from "react-native";
import { Dropdown } from "react-native-element-dropdown";
import type { FormInputRef } from "./FormInput";

type Option = {
  label: string;
  value: string;
};

type Props = {
  label: string;
  value?: string;
  options: Option[];
  onChange: (value: string) => void;
  error?: string;
  placeholder?: string;
  disabled?: boolean;
  onLayoutY?: (y: number) => void; // report Y like other inputs
};

const FormDropdown = React.forwardRef<FormInputRef, Props>(
  (
    {
      label,
      value,
      options,
      onChange,
      error,
      placeholder = "Select...",
      disabled,
      onLayoutY,
    },
    ref
  ) => {
    const dropdownRef = useRef<any>(null);

    useImperativeHandle(ref, () => ({
      focus: () => {
        if (dropdownRef.current?.focus) dropdownRef.current.focus();
        else if (dropdownRef.current?.open) dropdownRef.current.open();
      },
    }));

    return (
      <View
        style={styles.inputGroup}
        onLayout={(e) => onLayoutY?.(e.nativeEvent.layout.y)}
      >
        <Text style={styles.label}>{label}</Text>
        <Dropdown
          ref={dropdownRef as any}
          style={[
            styles.dropdown,
            error ? styles.inputError : styles.inputBorder,
          ]}
          placeholderStyle={styles.placeholderText}
          selectedTextStyle={styles.selectedText}
          itemTextStyle={styles.dropdownText}
          iconStyle={styles.dropdownIcon}
          containerStyle={styles.dropdownMenu}
          itemContainerStyle={styles.itemContainer}
          activeColor={secondary.dark}
          data={options}
          labelField="label"
          valueField="value"
          placeholder={placeholder}
          value={value}
          onChange={(item) => onChange(item.value)}
          disable={disabled}
        />
        {error && <Text style={styles.error}>{error}</Text>}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  inputGroup: {
    marginBottom: 18,
  },
  label: Platform.select({
    default: { fontSize: 15, marginBottom: 6, color: secondary.dark },
    android: { fontSize: 13, marginBottom: 4, color: secondary.dark },
  }),
  dropdown: Platform.select({
    default: {
      backgroundColor: "white",
      borderWidth: 1,
      borderRadius: 0,
      padding: 8,
      height: 40,
      width: "100%",
      borderColor: secondary.dark,
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "space-between",
    },
    android: {
      backgroundColor: "white",
      borderWidth: 1,
      borderRadius: 0,
      padding: 8,
      height: 36,
      width: "100%",
      borderColor: secondary.dark,
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "space-between",
    },
  }),
  placeholderText: {
    color: secondary.dark,
    fontSize: Platform.select({ default: 16, android: 14 }) as number,
  },
  selectedText: {
    color: secondary.dark,
    fontSize: Platform.select({ default: 16, android: 14 }) as number,
    backgroundColor: "white",
  },
  dropdownText: {
    color: "white",
    fontSize: Platform.select({ default: 16, android: 14 }) as number,
    backgroundColor: secondary.dark,
    borderWidth: 0,
    padding: 0,
  },
  itemContainer: {
    backgroundColor: secondary.dark,
    borderWidth: 0,
    padding: 0,
    borderBottomWidth: 1,
    borderBottomColor: "#444",
    color: "white",
    opacity: 1,
  },
  dropdownIcon: {
    color: secondary.dark,
    fontSize: Platform.select({ default: 16, android: 14 }) as number,
    paddingLeft: 8,
    borderWidth: 0,
  },
  dropdownMenu: {
    borderWidth: 1,
    borderColor: secondary.dark,
    borderRadius: 0,
    backgroundColor: secondary.dark,
  },
  inputBorder: {
    borderColor: secondary.dark,
  },
  inputError: {
    borderColor: errorColor,
  },
  error: {
    color: errorColor,
    fontSize: 14,
    marginTop: 6,
  },
});

export default FormDropdown;
