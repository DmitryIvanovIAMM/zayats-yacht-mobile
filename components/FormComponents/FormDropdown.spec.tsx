import { act, fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { FormProvider, useForm } from "react-hook-form";
import FormDropdown from "./FormDropdown";

// Mock element dropdown to a simple selectable stub
jest.mock("react-native-element-dropdown", () => {
  const React = require("react");
  const { View, Text, TouchableOpacity } = require("react-native");
  const Dropdown = (props: any) => (
    <View accessibilityLabel="mock-dropdown" {...props}>
      <Text>{props.placeholder}</Text>
      {props.disable ? <Text>DISABLED</Text> : null}
      <TouchableOpacity
        accessibilityLabel="select-first"
        onPress={() => props.onChange?.(props.data?.[0])}
      />
      <TouchableOpacity
        accessibilityLabel="select-second"
        onPress={() => props.onChange?.(props.data?.[1])}
      />
    </View>
  );
  return { __esModule: true, Dropdown };
});

type FormValues = { purpose: string };

function TestFormProvider({
  children,
  defaultValues,
  onReady,
}: {
  children: React.ReactNode;
  defaultValues?: Partial<FormValues>;
  onReady?: (methods: any) => void;
}) {
  const methods = useForm<FormValues>({
    defaultValues: { purpose: "", ...(defaultValues || {}) },
  });
  React.useEffect(() => {
    onReady?.(methods);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return <FormProvider {...methods}>{children}</FormProvider>;
}

function renderWithForm(
  ui: React.ReactElement,
  defaultValues?: Partial<FormValues>
) {
  let methodsRef: any;
  const utils = render(
    <TestFormProvider
      defaultValues={defaultValues}
      onReady={(m) => (methodsRef = m)}
    >
      {ui}
    </TestFormProvider>
  );
  return { ...utils, methods: methodsRef as any };
}

describe("FormDropdown", () => {
  const options = [
    { label: "One", value: "one" },
    { label: "Two", value: "two" },
  ];

  it("matches snapshot", () => {
    const { toJSON } = renderWithForm(
      <FormDropdown
        name="purpose"
        label="Purpose"
        options={options}
        placeholder="Select..."
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders label, placeholder, and shows disabled state", () => {
    const { getByText } = renderWithForm(
      <FormDropdown
        name="purpose"
        label="Purpose"
        options={options}
        placeholder="Select..."
        disabled
      />
    );

    expect(getByText("Purpose")).toBeTruthy();
    expect(getByText("Select...")).toBeTruthy();
    expect(getByText("DISABLED")).toBeTruthy();
  });

  it("updates form value when an option is selected", async () => {
    const { getByLabelText, methods } = renderWithForm(
      <FormDropdown
        name="purpose"
        label="Purpose"
        options={options}
        placeholder="Select..."
      />
    );

    await act(async () => {
      fireEvent.press(getByLabelText("select-first"));
    });
    await waitFor(() => expect(methods.getValues("purpose")).toBe("one"));

    await act(async () => {
      fireEvent.press(getByLabelText("select-second"));
    });
    await waitFor(() => expect(methods.getValues("purpose")).toBe("two"));
  });

  it("shows an error message from react-hook-form", async () => {
    const { getByText, methods } = renderWithForm(
      <FormDropdown
        name="purpose"
        label="Purpose"
        options={options}
        placeholder="Select..."
      />
    );

    await act(async () => {
      methods.setError("purpose", { message: "Please select a purpose" });
    });

    await waitFor(() =>
      expect(getByText("Please select a purpose")).toBeTruthy()
    );
  });
});
