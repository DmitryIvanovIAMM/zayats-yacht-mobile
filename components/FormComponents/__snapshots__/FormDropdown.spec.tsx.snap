// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDropdown matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "marginBottom": 8,
    }
  }
>
  <Text
    style={
      {
        "color": "#006666",
        "fontSize": 15,
        "marginBottom": 6,
      }
    }
  >
    Purpose
  </Text>
  <View
    accessibilityLabel="mock-dropdown"
    activeColor="#006666"
    containerStyle={
      {
        "backgroundColor": "#006666",
        "borderColor": "#006666",
        "borderRadius": 0,
        "borderWidth": 1,
      }
    }
    data={
      [
        {
          "label": "One",
          "value": "one",
        },
        {
          "label": "Two",
          "value": "two",
        },
      ]
    }
    iconStyle={
      {
        "borderWidth": 0,
        "color": "#006666",
        "fontSize": 16,
        "paddingLeft": 8,
      }
    }
    itemContainerStyle={
      {
        "backgroundColor": "#006666",
        "borderBottomColor": "#444",
        "borderBottomWidth": 1,
        "borderWidth": 0,
        "color": "white",
        "opacity": 1,
        "padding": 0,
      }
    }
    itemTextStyle={
      {
        "backgroundColor": "#006666",
        "borderWidth": 0,
        "color": "white",
        "fontSize": 16,
        "padding": 0,
      }
    }
    labelField="label"
    onChange={[Function]}
    placeholder="Select..."
    placeholderStyle={
      {
        "color": "#006666",
        "fontSize": 16,
      }
    }
    selectedTextStyle={
      {
        "backgroundColor": "white",
        "color": "#006666",
        "fontSize": 16,
      }
    }
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "#006666",
          "borderRadius": 0,
          "borderWidth": 1,
          "display": "flex",
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "space-between",
          "padding": 8,
          "width": "100%",
        },
        {
          "borderColor": "#006666",
        },
      ]
    }
    testID="mock-dropdown"
    value=""
    valueField="value"
  >
    <Text>
      Select...
    </Text>
    <View
      accessibilityLabel="select-first"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    />
    <View
      accessibilityLabel="select-second"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    />
  </View>
  <View
    style={
      {
        "justifyContent": "center",
        "marginTop": 2,
        "minHeight": 18,
      }
    }
  />
</View>
`;
